test/unit/go-parser-memory-stream.spec.ts

```diff
@@ -0,0 +1,174 @@
+
+        elf.parse = originalParse;
+      }
+    });
+  });
+
+  describe("ELF magic detection", () => {
+    it("should detect ELF magic correctly", async () => {
+      const elfContent = Buffer.concat([
+        Buffer.from("\x7FELF"), // ELF magic
+        Buffer.alloc(100, 0),
+      ]);
+      const stream = createStreamFromBuffer(elfContent);
+
+      // Mock elf.parse to avoid complexity
+      const originalParse = elf.parse;
+      elf.parse = jest.fn().mockReturnValue({ body: { sections: [] } });
+
+      const result = await findGoBinaries(stream, elfContent.length);
+
+      expect(elf.parse).toHaveBeenCalled();
+      expect(result).toBeUndefined(); // No Go sections
+
+      elf.parse = originalParse;
+    });
+
+    it("should reject non-ELF files early", async () => {
+      const nonElfContent = Buffer.from("Not ELF content");
+      const stream = createStreamFromBuffer(nonElfContent);
+
+      const result = await findGoBinaries(stream, 1024 * 1024); // 1MB
+
+      expect(result).toBeUndefined();
+    });
+
+    it("should reject ELF magic split across chunks (current limitation)", async () => {
+      // Our current implementation doesn't handle ELF magic split across chunks
+      // This is acceptable because it's extremely rare in real TAR streams
+      const stream = new Readable();
+      stream.push(Buffer.from("\x7F")); // First chunk - not full ELF magic
+      stream.push(Buffer.from("ELF")); // Second chunk
+      stream.push(Buffer.alloc(60, 0)); // Rest
+      stream.push(null);
+
+      const result = await findGoBinaries(stream);
+
+      // Should reject because first chunk doesn't contain full ELF magic
+      expect(result).toBeUndefined();
+    });
+  });
+
+  describe("filePathMatches function", () => {
+    it("should match files without extensions", () => {
+      expect(filePathMatches("/app/myservice")).toBe(true);
+      expect(filePathMatches("/usr/bin/kubectl")).toBe(true);
+      expect(filePathMatches("/opt/binary")).toBe(true);
+    });
+
+    it("should not match files with extensions", () => {
+      expect(filePathMatches("/app/script.sh")).toBe(false);
+      expect(filePathMatches("/app/config.json")).toBe(false);
+      expect(filePathMatches("/app/main.go")).toBe(false);
+    });
+
+    it("should not match files in ignored paths", () => {
+      expect(filePathMatches("/etc/passwd")).toBe(false);
+      expect(filePathMatches("/var/log/app")).toBe(false);
+      expect(filePathMatches("/tmp/file")).toBe(false);
+      expect(filePathMatches("/proc/cpuinfo")).toBe(false);
+    });
+
+    it("should handle the problematic case - large files without extensions", () => {
+      // These would previously cause the memory issue
+      expect(filePathMatches("/app/large-data-file")).toBe(true);
+      expect(filePathMatches("/usr/local/huge-binary")).toBe(true);
+    });
+  });
+
+  describe("Error handling", () => {
+    it("should handle stream errors gracefully", async () => {
```

Comment:
nit: throwing an error is graceful 🤨 ? What would non-graceful handling look like?